File: app/src/main/java/game/ivan/ashancalculator/tags/controller/TagStateView.java
package game.ivan.ashancalculator.tags.controller;

import com.hannesdorfmann.mosby.mvp.viewstate.ViewState;

import java.util.List;

import game.ivan.ashancalculator.database.models.Tags;
import game.ivan.ashancalculator.tags.view.TagsView;
import io.reactivex.Observable;

/**
 * Created by ivan on 05.01.2017.
 */

public class TagStateView implements ViewState<TagsView> {
    public final static int SHOW_CONTENT=1;
    public final static int SHOW_CREATE_DIALOG = 2;
    public final static int SHOW_EDIT_DIALOG = 3;

    int currentViewState=1;

    public void setTags(List<Tags> tags) {
        this.tags = tags;
    }

    List<Tags> tags;

    public void setEditableTag(Tags editableTag) {
        this.editableTag = editableTag;
    }

    Tags editableTag;

    @Override
    public void apply(TagsView view, boolean retained) {
        switch (currentViewState){
            case SHOW_CONTENT:
                view.refreshList(Observable.just(tags));
                break;
            case SHOW_CREATE_DIALOG:
                view.refreshList(Observable.just(tags));
                view.showCreateDialog();
                break;
            case SHOW_EDIT_DIALOG:
                view.refreshList(Observable.just(tags));
                view.showEditDialog(editableTag);
                break;
        }
    }

    public void setShowContentState(){
        currentViewState=SHOW_CONTENT;
    }

    public void setShowCreateDialogState(){
        currentViewState=SHOW_CREATE_DIALOG;
    }

    public void setShowEditDialogState(){
        currentViewState=SHOW_EDIT_DIALOG;
    }
}

File: app/src/main/java/game/ivan/ashancalculator/tags/controller/TagStateView.kt
package game.ivan.ashancalculator.tags.controller

import com.hannesdorfmann.mosby.mvp.viewstate.ViewState
import game.ivan.ashancalculator.database.models.Tags
import game.ivan.ashancalculator.tags.view.TagsView

/**
 * Created by ivan on 05.02.2017.
 */
class TagStateView: ViewState<TagsView>{


    companion object{
        @JvmStatic val  SHOW_CONTENT:Int = 1
        @JvmStatic val  SHOW_CREATE_DIALOG:Int = 2
        @JvmStatic val  SHOW_EDIT_DIALOG:Int = 3
    }

    var currentViewState:Int = 1
    lateinit var tages:List<Tags>
    lateinit var editTags:Tags

    fun setTags(list:List<Tags>){
        tages = list
    }

    fun setEditableTag(tag:Tags){
        editTags = tag
    }

    override fun apply(view: TagsView?, retained: Boolean) {
        when(currentViewState){
            SHOW_CONTENT ->  view!!.refreshList(tages)
            SHOW_CREATE_DIALOG ->  {
                view!!.refreshList(tages)
                view!!.showCreateDialog()}
            SHOW_EDIT_DIALOG ->  {
                view!!.refreshList(tages)
                view!!.showEditDialog(editTags)
            }
        }
    }

    fun setShowContentState(){
        currentViewState=SHOW_CONTENT
    }

    fun setShowCreateDialogState(){
        currentViewState=SHOW_CREATE_DIALOG
    }

    fun setShowEditDialogState(){
        currentViewState=SHOW_EDIT_DIALOG
    }

}